rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp

```diff
@@ -698,18 +640,19 @@ class RMW_Connext_SubscriberStatusCondition : public RMW_Connext_StatusCondition
   inline rmw_ret_t
   get_liveliness_changed_status(rmw_liveliness_changed_status_t * const status)
   {
-    std::lock_guard<std::mutex> lock(this->mutex_internal);
+    update_state(
+      [this, status]() {
+        this->triggered_liveliness = false;
 
-    this->triggered_liveliness = false;
+        status->alive_count = this->status_liveliness.alive_count;
+        status->alive_count_change = this->status_liveliness.alive_count_change;
+        status->not_alive_count = this->status_liveliness.not_alive_count;
+        status->not_alive_count_change = this->status_liveliness.not_alive_count_change;
 
-    status->alive_count = this->status_liveliness.alive_count;
-    status->alive_count_change = this->status_liveliness.alive_count_change;
-    status->not_alive_count = this->status_liveliness.not_alive_count;
-    status->not_alive_count_change = this->status_liveliness.not_alive_count_change;
-
-    this->status_liveliness.alive_count_change = 0;
-    this->status_liveliness.not_alive_count_change = 0;
-    this->status_liveliness_last = this->status_liveliness;
+        this->status_liveliness.alive_count_change = 0;
+        this->status_liveliness.not_alive_count_change = 0;
+        this->status_liveliness_last = this->status_liveliness;
```

Comment:
out of curiosity, can't `alive_count` and `alive_count_change` be copied directly from the DDS internal instead of calculated with `status_liveliness_last` and `status_liveliness`?
I don't quite follow why that's needed, though it's also not wrong.

Reply:
The "change" fields must report the change in the other counters from the last time the status was reported to the application. Since the RMW installs listeners on writers and readers, the middleware considers a listener notification as "having reported the status", so the status changes reported to the listeners and cached by the RMW cannot be propagated directly to the application, but rather the RMW must keep track of the last value seen by the application, and basically re-implement what is already done by the middleware. It would be great if there was a way for the listener to not reset the status implicitly, so that these `get_*()` functions could rely on the standard DDS `get_status()` functions, but alas that is not possible at the moment.

Reply:
makes sense, thanks for explaining!